### ISSUE TITLE
Persisting plist updating keys using symbols not working, while string constants work fine

### ISSUE THREAD

**Issue by zats** [fastlane/fastlane#5905]
I'm using `update_info_plist` action like so:

```rb
update_info_plist(plist_path: INFO_PLIST, xcodeproj: XCODEPROJ_PATH, block: lambda { |plist|
  plist['working'] = 'foo'
  plist[:not_working] = 'bar'
}
```

First change is being persisted to the file, second change is not. 
It appears there is a bug in [`Xcodeproj::Plist.write_to_path`](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/update_info_plist.rb#L45).

I'm using rvm with ruby 2.2.2
Please let me know if I can provide any more information to solve this.

**Comment by TKBurner** [comment]
@zats Hope you are doing well! I wanted to see if you can share the full input and output here? Are you getting any errors or is the plist just not being updated? Thanks!

**Comment by zats** [comment]
Hi Todd, how are you? 
Sorry for not being clear, I'm not getting any errors, the action reports successful completion, it's just the values assigned using `:symbols` notation are not being persisted to the disk.

**Comment by TKBurner** [comment]
Hi @zats Thanks for following up. Any chance you can provide a sample project or exact repro steps. I want to circle in the core team. Thanks!

**Comment by zats** [comment]
Hi Todd, here you go, following is the relevant part of the fastfile

```rb
platform :ios do
# ...
  lane :beta do |options|
    update_info_plist(plist_path: INFO_PLIST, xcodeproj: XCODEPROJ_PATH, block: lambda { |plist|
      # ...
      plist[:CFBundleDisplayName] = 'Wondermall'
      plist['FacebookUrlSchemeSuffix'] = ''
      # ...
    })
  end
end
```

This is a relevant part. As I mentioned above, the only modification we saw in the plist is `FacebookUrlSchemeSuffix` key is being set to an empty string, value for `CFBundleDisplayName` is  not persisted in the plist

**Comment by TKBurner** [comment]
Thanks @zats! I looped in the core team to see if they can take a look. Thanks!

**Comment by frisocki** [comment]
When the existing plist file with the key CFBundleDisplayName is read, it is stored in the hash as the string literal `CFBundleDisplayName`. When you attempt to update that value with the symbol `:CFBundleDisplayName`, 2 entries are created in the hash:

1. String Literal CFBundleDisplayName
2. Symbol :CFBundleDisplayName

Then when the plist file is written, it attempts to write out both keys and values. Each symbol must be converted to a string, at which point there are basically duplicate keys in the hash.

The underlying implementation ultimately calls (in projects outside of fastlane)

[RubyHashToCFDictionary](http://www.rubydoc.info/github/CocoaPods/Xcodeproj/CoreFoundation.RubyHashToCFDictionary)

```ruby
hash.each do |key, value|
    key = RubyStringToCFString(key.to_s)
    value = RubyValueToCFTypeRef(value)
    CFDictionaryAddValue(result, key, value)
  end
```

[CFDictionaryAddValue](https://developer.apple.com/library/ios/documentation/CoreFoundation/Reference/CFMutableDictionaryRef/#//apple_ref/c/func/CFDictionaryAddValue)

> Adds a key-value pair to a dictionary if the specified key is not already present.

The 1st `CFDictionaryAddValue` will work, subject to how `hash.each` decides to enumerate the keys and values. The 2nd will silently fail.

Any reason why you are using a symbol as the hash key and not the string literal? Using a string literal should solve your issue.

The reason it works the 1st time around is there is no collision in the hash key naming.

**Comment by mfurtak** [comment]
Hi @zats - thanks for the report 👍 

If there is indeed a problem with `Xcodeproj::Plist.write_to_path`, that would probably be best to raise with [CocoaPods/Xcodeproj](https://github.com/CocoaPods/Xcodeproj/).

That said, so far I have not been able to reproduce your experience...  Can you run

```
gem which fastlane ffi plist xcodeproj
```

And paste us the output? Thanks!

**Comment by zats** [comment]
Thank you everyone for response

@frisocki's suggestion does make sense and that is probably the case that I had values for aforementioned keys previously set, saying that, I assumed (probably wrongly) that once read into memory plist will coalesce keys whether they are referenced though a symbol or a string.

@mfurtak that's a good point. Definitely doesn't look like a bug in fastlane 
Here is the output of the command:
`/usr/local/lib/ruby/gems/2.3.0/gems/xcodeproj-1.2.0/lib/xcodeproj.rb`

**Comment by mfurtak** [comment]
Hm... odd. Is that all of the output? I'd expect to see lines for `fastlane`, `ffi`, and plist` as well.

**Comment by zats** [comment]
Oh, I'm sorry, I didn't copy all of it, here's the full output:
```
/usr/local/lib/ruby/gems/2.3.0/gems/fastlane-1.102.0/lib/fastlane.rb
ERROR:  Can't find ruby library file or shared library ffi
/usr/local/lib/ruby/gems/2.3.0/gems/plist-3.1.0/lib/plist.rb
/usr/local/lib/ruby/gems/2.3.0/gems/xcodeproj-1.2.0/lib/xcodeproj.rb
```
doesn't seem like I have `ffi` at all?

**Comment by mfurtak** [comment]
OK, that may be an interesting hint as to why I can't reproduce.

If I'm reading their code correctly, `Xcodeproj` internally chooses between two possible implementations for writing plist data. One attempts to call `CoreFoundation` and `DevToolsCore` code directly through FFI, and the other uses another gem called `plist`.

If you're still interested in experimenting, I might suggest trying to `gem install ffi` and then re-run your tests to see if the results are different. If so, that might suggest that the two transitive plist writing implementations have differences.

In either case, is this something that is still blocking you? Does the answer "use string keys for maximum safety" seem like a reasonable resolution for now? If so, are you aware of any documentation we should update for correctness?

**Comment by zats** [comment]
Oh yes we fixed the issue by replacing everything with strings. But I don't
mind to experiment to get to the bottom of this. Will try to do it shortly

**Comment by zats** [comment]
Interesting enough, even with `ffi` installed, it doesn't quite work for me
Here is the latest output of the `$ gem which fastlane ffi plist xcodeproj `

```
/usr/local/lib/ruby/gems/2.3.0/gems/fastlane-1.102.0/lib/fastlane.rb
/usr/local/lib/ruby/gems/2.3.0/gems/ffi-1.9.14/lib/ffi.rb
/usr/local/lib/ruby/gems/2.3.0/gems/plist-3.1.0/lib/plist.rb
/usr/local/lib/ruby/gems/2.3.0/gems/xcodeproj-1.2.0/lib/xcodeproj.rb
```

I set up a test project here https://github.com/zats/plist-bug
Sadly, when I was creating a test Xcode project, creativity left me and I called it `fastlane`, so don't be surprised to find both Xcode project files and the `Fastfile` in the same folder.

Basically there are two lanes, both modify just the plist, one of them works, another doesn't .
Let me know if it works on your machine?

Thanks!

**Comment by mfurtak** [comment]
Thanks for the example project, that's 💯 ! I will probably get to have a look tomorrow.

**Comment by mfurtak** [comment]
Using your project I was able to reproduce the problem. Thank you! 👍 

Using a symbol as the key will work, but only if that key is **not present** in the Plist at the time it is read. So, effectively, it will only work the first time 😫  All the reasons are exactly as @frisocki said above.

Because this problem has a simple work-around, I think I'm going to leave this here.

If we wanted to provide a little more safety, instead of [passing the hash to the block directly](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/update_info_plist.rb#L42), we could pass something that proxied for the hash and called `to_s` on all provided keys before updating the hash. If you're interested in making such a PR, I'd be happy to review it!

Thanks!
